diff -uNr vanilla-source-4.19/arch/arm/boot/dts/kirkwood-linkstation-6282.dtsi linux-source-4.19/arch/arm/boot/dts/kirkwood-linkstation-6282.dtsi
--- vanilla-source-4.19/arch/arm/boot/dts/kirkwood-linkstation-6282.dtsi	2019-11-10 03:08:58.335469083 -0600
+++ linux-source-4.19/arch/arm/boot/dts/kirkwood-linkstation-6282.dtsi	2019-11-10 03:27:44.062249159 -0600
@@ -115,8 +115,8 @@
 		pinctrl-0 = <&pmx_fan_low &pmx_fan_high &pmx_fan_lock>;
 		pinctrl-names = "default";
 
-		gpios = <&gpio0 17 GPIO_ACTIVE_LOW
-			 &gpio0 16 GPIO_ACTIVE_LOW>;
+		gpios = <&gpio0 17 GPIO_ACTIVE_HIGH
+			 &gpio0 16 GPIO_ACTIVE_HIGH>;
 
 		gpio-fan,speed-map = <0 3
 				1500 2
diff -uNr vanilla-source-4.19/arch/arm/boot/dts/kirkwood-lsxl.dtsi linux-source-4.19/arch/arm/boot/dts/kirkwood-lsxl.dtsi
--- vanilla-source-4.19/arch/arm/boot/dts/kirkwood-lsxl.dtsi	2019-11-10 03:08:58.335469083 -0600
+++ linux-source-4.19/arch/arm/boot/dts/kirkwood-lsxl.dtsi	2019-11-10 03:24:11.805494271 -0600
@@ -165,8 +165,8 @@
 		compatible = "gpio-fan";
 		pinctrl-0 = <&pmx_fan_low &pmx_fan_high &pmx_fan_lock>;
 		pinctrl-names = "default";
-		gpios = <&gpio0 19 GPIO_ACTIVE_LOW
-		         &gpio0 18 GPIO_ACTIVE_LOW>;
+		gpios = <&gpio0 19 GPIO_ACTIVE_HIGH
+		         &gpio0 18 GPIO_ACTIVE_HIGH>;
 		gpio-fan,speed-map = <0    3
 		                      1500 2
 		                      3250 1
