diff -uNr vanilla-source-4.9/arch/arm/mach-mv78xx0/buffalo-wxl-setup.c linux-source-4.9/arch/arm/mach-mv78xx0/buffalo-wxl-setup.c
--- vanilla-source-4.9/arch/arm/mach-mv78xx0/buffalo-wxl-setup.c	2019-08-22 16:19:25.076116508 -0500
+++ linux-source-4.9/arch/arm/mach-mv78xx0/buffalo-wxl-setup.c	2019-08-22 16:23:32.060763209 -0500
@@ -128,16 +128,8 @@
 
 static int __init wxl_pci_init(void)
 {
-	if (machine_is_terastation_wxl()) {
-		/*
-		 * Assign the x16 PCIe slot on the board to CPU core
-		 * #0, and let CPU core #1 have the four x1 slots.
-		 */
-		if (mv78xx0_core_index() == 0)
-			mv78xx0_pcie_init(0, 1);
-		else
-			mv78xx0_pcie_init(1, 0);
-	}
+	if (machine_is_terastation_wxl() && mv78xx0_core_index() == 0)
+                mv78xx0_pcie_init(1, 1);
 
 	return 0;
 }
diff -uNr vanilla-source-4.9/arch/arm/mach-mv78xx0/pcie.c linux-source-4.9/arch/arm/mach-mv78xx0/pcie.c
--- vanilla-source-4.9/arch/arm/mach-mv78xx0/pcie.c	2019-08-22 16:19:25.076116508 -0500
+++ linux-source-4.9/arch/arm/mach-mv78xx0/pcie.c	2019-08-22 16:21:57.834042744 -0500
@@ -45,7 +45,7 @@
 
 u32 pcie_port_size[8] = {
 	0,
-	0x30000000,
+	0x20000000,
 	0x10000000,
 	0x10000000,
 	0x08000000,
